### PR TITLE
Feature/separate window and form

### LINKED ISSUE #23: Allow child elements to have custom names (should be fixed in Core lib)

### LINKED ISSUE #36: Separate Window and Form page objects
Search Window pageobjects from the root sessions and Form pageobjects - from the parent window context or the application session context
Use respective WindowsDriver session for each element

### PR DESCRIPTION
**Pull request by mialeska**
Closes #23 and closes #36

### CODE REVIEW

File: Aquality.WinAppDriver/src/Aquality.WinAppDriver/Applications/Application.cs [hunk 1]
```diff
@@ -70,12 +70,16 @@ public WindowsDriver<WindowsElement> RootSession
             }
         }
 
+        public virtual bool IsStarted => IsSessionStarted(applicationSession) || IsSessionStarted(rootSession);
+
+        private bool IsSessionStarted(WindowsDriver session) => session != null && session.SessionId != null;
```
Comment (knysh) [inline Aquality.WinAppDriver/src/Aquality.WinAppDriver/Applications/Application.cs]: you could update it like session?.SessionId != null

File: Aquality.WinAppDriver/src/Aquality.WinAppDriver/Forms/Window.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+﻿using OpenQA.Selenium;
+
+namespace Aquality.WinAppDriver.Forms
+{
+    /// <summary>
+    /// Defines base class for a separate window of any application.
+    /// </summary>
+    public abstract class Window : Form
```
Comment (knysh) [inline Aquality.WinAppDriver/src/Aquality.WinAppDriver/Forms/Window.cs:10]: I suppose, that Window as biggest entity should be parent of Form and Form should be parent for element, could you clarify why did you  со? 
e.g. I not sure that window should have ParentForm

File: Aquality.WinAppDriver/src/Aquality.WinAppDriver/Elements/ElementFactory.cs [hunk 3]
```diff
@@ -20,29 +21,40 @@ namespace Aquality.WinAppDriver.Elements
...
+            Func<ISearchContext> searchContextSupplier = null, 
+            bool isRootSession = false) 
+            : base(conditionalWait, elementFinder, localizationManager)
         {
+            this.searchContextSupplier = searchContextSupplier;
+            this.isRootSession = isRootSession;
         }
 
-        public IButton GetButton(By locator, string name, ElementState elementState = ElementState.Displayed)
+        public virtual IButton GetButton(By locator, string name, ElementState elementState = ElementState.Displayed)
```
Comment (knysh) [inline Aquality.WinAppDriver/src/Aquality.WinAppDriver/Elements/ElementFactory.cs:37]: Are you sure we need this methods as virtual ?

File: Aquality.WinAppDriver/src/Aquality.WinAppDriver/Forms/Form.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        /// Constructor with parameters.
+        /// </summary>
+        /// <param name="locator">Unique locator of the window.</param>
+        /// <param name="name">Name of the window.</param>
+        /// <param name="parentForm">Parent form. If set to null, search context of <see cref="AqualityServices.Application"/> is used.</param>
+        /// <param name="isRootSession">Determines whether the search of the current form would be performed from the <see cref="IWindowsApplication.RootSession"/> or not.
+        /// If is set to false, search is performed from the application session <see cref="IWindowsApplication.Driver"/>.</param>
+        /// <param name="elementState">Element presence state.</param>
+        protected Form(By locator, string name, IForm parentForm = null, bool isRootSession = false, ElementState elementState = ElementState.Displayed)
+            : base(locator, name, ResolveSearchContextSupplier(parentForm), parentForm == null ? isRootSession : parentForm.IsRootSession, elementState)
```
Comment (knysh) [inline Aquality.WinAppDriver/src/Aquality.WinAppDriver/Forms/Form.cs]: parentForm?.IsRootSession

File: Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementFactoryTests.cs [hunk 5]
```diff
@@ -10,7 +10,7 @@ namespace Aquality.WinAppDriver.Tests.Elements
 {
     public class ElementFactoryTests : TestWithApplication
     {
-        private static readonly CalculatorWindow calculatorWindow = new CalculatorWindow();
+        private static readonly CalculatorForm calculatorWindow = new CalculatorForm();
```
Comment (knysh) [inline Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementFactoryTests.cs]: calculatorForm?

File: Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs [hunk 6]
```diff
@@ -11,7 +11,7 @@ namespace Aquality.WinAppDriver.Tests.Elements
 {
     public class ElementStateProviderTests : TestWithApplication
     {
-        private static readonly CalculatorWindow CalculatorWindow = new CalculatorWindow();
+        private static readonly CalculatorForm CalculatorWindow = new CalculatorForm();
```
Comment (knysh) [inline Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs]: CalculatorWindow  -> CalculatorForm

File: Aquality.WinAppDriver/src/Aquality.WinAppDriver/Elements/ElementFactory.cs (hunk 3 above)
Comment (mialeska) [reply Aquality.WinAppDriver/src/Aquality.WinAppDriver/Elements/ElementFactory.cs:37]: yes, for the case of custom buttons on the project where this library is used

File: Aquality.WinAppDriver/src/Aquality.WinAppDriver/Forms/Form.cs (hunk 4 above)
Comment (mialeska) [reply Aquality.WinAppDriver/src/Aquality.WinAppDriver/Forms/Form.cs]: it won't return the same result, e.g. when parentForm is null and isRootSession is true I expect to get the "true" here

File: Aquality.WinAppDriver/src/Aquality.WinAppDriver/Forms/Window.cs (hunk 2 above)
Comment (mialeska) [reply Aquality.WinAppDriver/src/Aquality.WinAppDriver/Forms/Window.cs:10]: Inheritance is not always the child-parent relationship.
From my perspective, 
a Form is an element with some nested elements; a Form could exist inside the separate window or inside the main application window; a Form could contain nested forms.
a Window is a Form which is not attached to the main application, and therefore should be found from the root session; a Window could contain nested elements and forms, but not the nested windows.

Regarding the ParentForm field, I cannot find any use for it and suggest to remove it.

File: Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementFactoryTests.cs (hunk 5 above)
Comment (mialeska) [reply Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementFactoryTests.cs]: thanks, fixed

File: Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs (hunk 6 above)
Comment (mialeska) [reply Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs]: thanks, fixed

File: Aquality.WinAppDriver/src/Aquality.WinAppDriver/Applications/Application.cs (hunk 1 above)
Comment (mialeska) [reply Aquality.WinAppDriver/src/Aquality.WinAppDriver/Applications/Application.cs]: seems that yes, thx

**Pull request closed without merging**